types/wlr_input_method_v2.c

```diff
@@ -20,10 +24,24 @@ static struct wlr_input_method_v2 *input_method_from_resource(
 	return wl_resource_get_user_data(resource);
 }
 
+static struct wlr_input_method_keyboard_grab_v2 *keyboard_grab_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_input_method_keyboard_grab_v2_interface,
+		&keyboard_grab_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void keyboard_grab_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab);
+
 static void input_method_destroy(struct wlr_input_method_v2 *input_method) {
 	wlr_signal_emit_safe(&input_method->events.destroy, input_method);
 	wl_list_remove(wl_resource_get_link(input_method->resource));
 	wl_list_remove(&input_method->seat_destroy.link);
+	if (input_method->keyboard_grab != NULL) {
+		keyboard_grab_destroy(input_method->keyboard_grab);
```

Comment:
wlroots destructors are no-op when called with NULL. We should move this NULL check in `keyboard_grab_destroy`.

Reply:
Fixed.